tests/unit/lms/resources/lti_launch_test.py

```diff
@@ -68,28 +70,70 @@ def with_canvas(self, pyramid_request):
         pyramid_request.product.family = Product.Family.CANVAS
 
 
+class TestCourse:
+    def test_it_when_existing(self, course_service, lti_launch):
+        course_service.get_by_context_id.return_value = factories.Course(
+            extra={"existing": "extra"}
+        )
+
+        course = lti_launch.course
+
+        course_service.get_by_context_id.assert_called_once_with(
+            sentinel.context_id,
+        )
+        course_service.upsert_course.assert_called_once_with(
+            context_id=sentinel.context_id,
+            name=sentinel.context_title,
+            extra={"existing": "extra"},
+        )
+        assert course == course_service.upsert_course.return_value
+
+    def test_it_when_new(self, course_service, lti_launch, _new_course_extra):
+        course_service.get_by_context_id.return_value = None
+
+        course = lti_launch.course
+
+        course_service.get_by_context_id.assert_called_once_with(sentinel.context_id)
+        course_service.upsert_course.assert_called_once_with(
+            context_id=sentinel.context_id,
+            name=sentinel.context_title,
+            extra=_new_course_extra.return_value,
+        )
+        assert course == course_service.upsert_course.return_value
+
+    @pytest.fixture
+    def _new_course_extra(self, lti_launch):
+        with patch.object(lti_launch, "_new_course_extra", autospec=True) as patched:
+            yield patched
+
+
 class TestGroupingType:
     def test_it(
         self,
         grouping_service,
         lti_launch,
-        course_service,
         assignment_service,
         pyramid_request,
+        course,
     ):
         assert (
             lti_launch.grouping_type
             == grouping_service.get_launch_grouping_type.return_value
         )
         assignment_service.get_assignment.assert_called_once_with(
-            mock.sentinel.tool_guid, mock.sentinel.resource_link_id
+            sentinel.tool_guid, sentinel.resource_link_id
         )
         grouping_service.get_launch_grouping_type.assert_called_once_with(
             pyramid_request,
-            course_service.upsert_course.return_value,
+            course,
```

Comment:
```suggestion
            lti_launch.course,
```

Alternatively we could use `lti_launch.course` and let it work out as it will? This is effectively the same command and would very closely follow the layout in the code without having to use a separate fixture.